Javascript Core 3 - Re-Write (#384)
Code-Your-Future/syllabus: js-core-3/week-2/lesson.md

# JavaScript Core III - Week 2

## Learning Objectives

- The learner should understand what the acronym API means

> - The learner can define what an `REST API`s purpose is and why it is useful

Reviewer: ```suggestion
- The learner can define what an `REST API`'s purpose is and why it is useful
```

> - The learner should understand the structure of `REST API`s url and how it can be manipulated to change the data

Reviewer: ```suggestion
- The learner should understand the structure of a `REST API` URL and how it can be manipulated to change the data
```

- The learner should understand what `fetch` is and what it is used for
- The learner should be able to use `fetch` to retrieve `JSON` from an `REST API`
- The learner should be able to parse the `JSON` and extract data from it
- The learner should be able to use `DOM` manipulation to add content to the `DOM`
- The learner should understand `window.onload` and `document.onload` and should be able to assign functions to run at specific life cycle events

## Agenda

The purpose of this class is to introduce to the student:

1. What are `APIs` and how to interact with them
2. How to use the `fetch` API to do AJAX calls

## 1. What are APIs and how to interact with them

### Explanation

- APIs are created by providers and used by consumers (BE provider, FE consumer)
- Part of an application that can be communicated with from an outside source
- Connect to it using "endpoints"
- Software well-known APIs (Fb APIs, Twitter APIs, Maps APIs, weather APIs);
- API doesn't care which language or technology is used in the consumer or the provider

#### Types of APIs:

- Private: for employees only under a company network for internal use.
- Semi-private: for clients who paid for the API.

> - Public: for everyone on the web.

Reviewer: ```suggestion
- Public: for everyone on the web (but may or may not need an account to use).
```

#### Architecture styles of API:

- Single purpose: API that gives a direct and specific service.
- Aggregated API: one API as a wrapper for multiple APIs.

> - Web services API: punch of APIs working together to forma whole app.

Reviewer: ```suggestion
- Web services API: bunch of APIs working together to form a whole app.
```

**Question (5 mins):**

    Give an example of a company that uses an API to allow access to their data.

#### Basic structure of REST API

- Endpoint: https://api.example.com
- Endpoint with version: https://api.example.com/v1
- Resources:

* https://api.example.com/v1/users
* https://api.example.com/v1/users/create
* https://api.example.com/v1/users/1
* https://api.example.com/v1/users/1/edit

- Query params:

* https://api.example.com/v1/users?limit=10

### Example

- Give real life example like (Devices like TV, any machine + electricity power socket interface which provides power to any external device)
- Give an example of an API in production
  - [Game Of Thrones API](https://api.tvmaze.com/singlesearch/shows?q=game%20of%20thrones)

### Exercise

#### Check-in

Here is an example of Spotify's REST endpoint to allow you to get the tracks from a playlist

    GET https://api.spotify.com/v1/playlists/{playlist_id}/tracks

**Task (5 mins):**

    On Instagram, you can access a post via the following endpoint:

    https://api.instagram.com/v1/media/{media-id}

    1. Write the URL of the endpoint to get the list of comments on a given instagram post?

>     2. What HTTP verb will you use to make requests to this endpoint?

Reviewer: I don't think we've taught them about HTTP verbs at this point?

> ## 2. How to use the `fetch` API to do AJAX calls

Reviewer: It's probably confusing to refer to `fetch` as an API here, given what we've introduced APIs to mean.

I'd probably just rephrase this as "How to use the `fetch` function to call get data from APIs"? We _could_ try to explain that some APIs are in-process / code, and others are out of process / need HTTP, but that seems overly complicated here.

50:50 on whether to mention AJAX - maybe we could have a footnote at the end of the topic saying "fetch used to be called XmlHttpRequest, and also AJAX - if you see those terms, they probably mean fetch" or something similar. The terms aren't crucial to understand `fetch`, so I'd take them off the critical path of learning, and just mention at the end "For historical reasons, you may see fetch referred to as AJAX, or as an XmlHttpRequest".

### Explanation

> - Modern replacement of XMLHttpRequest

Reviewer: If they haven't seen XHR before, I probably wouldn't mention it here.

- Uses Promise structure
- The Fetch API is defined in the browser (window.fetch)
- Only modern browsers support it (show [caniuse.com](https://caniuse.com/#feat=fetch))
- Fetch API documentations by mozilla [link](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch)

> ### What is Fetch made of?

Reviewer: Can we introduce promises _before_ fetch? This feels like a detour, and I worry that:
1. It will make people think that promises are _only_ used for fetch
2. It'll make it harder to grasp fetch (because people should be thinking HTTP, and suddenly they're not)

> Fetch is an example of a javascript which has a very specific structure.

Reviewer: ```suggestion
Fetch uses a javascript pattern called "Promises" which has a very specific structure.

Using Promises allows us to schedule functions to be called after some asynchronous code finishes running. We can specify different functions depending on whether the asynchronous code was successful or ran into an error.

Promises can make it easier to split our code into small well-named functions, and make code easier to read. They also make it easier to handle errors.
```

```javascript

> let promiseToDoHomeWork = new Promise(function (resolve, reject) {

Reviewer: How about a slightly less artificial example?
```javascript
let clickCount;

document.querySelector("button").addEventListener("click", () => clickCount++);

let promiseOfClicks = new Promise(function(resolve, reject) {
  clickCount = 0;
  console.log("Counting clicks for 5 seconds...");
  setTimeout(() => {
    if (clickCount > 0) {
      resolve(clickCount);
    } else {
      reject();
    }
  }, 5000)
});

promiseOfClicks
  .then(clicks => console.log(`Clicked ${clicks} times`)
  .catch(error => console.warn(error));
```

It's still pretty artificial, but at least it passes around _some_ data...

I worry that when we use entirely contrived examples, the students often don't get the point...

    console.warn(err);
  });
```

### Example

#### Example 1

```javascript
//Retrieve the JSON
fetch("https://seriousnews.com/api/headlines")
  // Get the response and extract the JSON
  .then(function (response) {
    return response.json();
  })
  // Do something with the JSON
  .then((headlines) => {
    console.log(headlines);
  })
  // If something goes wrong
  .catch((error) => console.log(error));
```

#### Example 2

Wouldn't it cool to make a new friend with just the click of a button?

Write a function that makes an API call using `fetch` to `https://www.randomuser.me/api`

- The function should make an API call to the given endpoint: `https://www.randomuser.me/api`
- Log the received data to the console
- Incorporate error handling
- Show how you can build a profile page for the user using the DOM
  - Add a name
  - Add a profile picture
  - Add some styling using CSS

### Exercises

#### Exercise 1

In groups the students should create a page of details about the United Kingdom.

The API endpoint can be found [here](https://restcountries.eu/rest/v2/name/Great%20Britain?fullText=true)

The website should include

- The name of the country
- The country's capital city

> - An unordered list of all the countries name in other languages

Reviewer: ```suggestion
- An unordered list of the country's name in other all of the other returned languages
```

##### Steps

Example `html` and `javascript` files can be found in the section below

1. Create a `HTML`, `CSS` and `JavaScript` file to hold different types of code
2. In your `HTML` file, write a simple basis for your website (e.g. [this](https://www.sitepoint.com/a-basic-html5-template/))
   - Make sure all of your `HTML`, `CSS` and `JavaScript` files are linked together!
3. Write a function using fetch that retrieves the `JSON` from the _Country API_
   - To make sure it's working print the JSON to the console using `console.log()`

> 4. Create a `h1` tag on the website using DOM manipulation and add the countries name inside it

Reviewer: ```suggestion
4. Create a `h1` tag on the website using DOM manipulation and add the country's name inside it
```

   - Go back to [Week 5](../../js-core-2/week-2/lesson.html) if you need a reminder

> 5. Create a `h2` tag on the website using DOM manipulation and add the capital cities name inside it

Reviewer: ```suggestion
5. Create a `h2` tag on the website using DOM manipulation and add the capital city's name inside it
```

6. Create a `ul` tag on the website using DOM manipulation
   - For each of the translated names in the JSON, add a `li` tag
7. Uncomment the lines inside `onLoad()` to load other countries details!

**Extra**

> - Load the countries flag into an `img` tag

Reviewer: ```suggestion
- Load the country's flag into an `img` tag
```

function addCountryCapital(countryData) {}

// This function should take the JSON for the country and put UL and LI tags on the screen with the countries name translated into other languages
function addNameInOtherLanguages(countryData) {}

function getContentDiv() {
  return document.querySelector("#content");
}

function onLoad() {
  getData(
    "https://restcountries.eu/rest/v2/name/Great%20Britain?fullText=true"
  );

  /** Remove this line when you have completed the task

  getData("https://restcountries.eu/rest/v2/name/France?fullText=true");

  getData("https://restcountries.eu/rest/v2/name/Germany?fullText=true");

  getData("https://restcountries.eu/rest/v2/name/Spain?fullText=true");

  getData("https://restcountries.eu/rest/v2/name/Portugal?fullText=true");

  getData("https://restcountries.eu/rest/v2/name/Hungary?fullText=true");

  getData("https://restcountries.eu/rest/v2/name/Russia?fullText=true");

  */ // Remove this line when you have completed the task
}

window.onload = onLoad;
```

##### Finished Example

You can find the finished example of this website [here](js-core-3\week-1\completed_country_website).

##### Check-in

**Question (5 mins):**

>     Which of the following statements about Promises is false?

Reviewer: I don't think this syllabus has covered enough about promises for the students to be able to answer these questions... The lesson may, unscripted, but...
